arangod/IResearch/IResearchDataStore.cpp

```diff
@@ -1425,13 +1451,23 @@ Result IResearchDataStore::initDataStore(
                          index().id().id(), "'")};
   }
 
-  if (!_dataStore._reader) {
+  if (!_dataStore.loadSnapshot()) {
     _dataStore._writer->commit();  // initialize 'store'
-    _dataStore._reader = irs::directory_reader::open(*(_dataStore._directory),
-                                                     _format, readerOptions);
+    auto engineSnapshot = _engine->currentSnapshot();
+    if (ADB_UNLIKELY(!engineSnapshot)) {
+      return {TRI_ERROR_INTERNAL,
+              absl::StrCat("Failed to get engine snapshot while initializing "
+                           "ArangoSearch index '",
+                           index().id().id(), "'")};
+    }
+    _dataStore.storeSnapshot(std::make_shared<DataSnapshot>(
+        irs::directory_reader::open(*(_dataStore._directory), _format,
+                                    readerOptions),
+        engineSnapshot));
   }
 
-  if (!_dataStore._reader) {
+  auto snapshot = _dataStore.loadSnapshot();
+  if (!snapshot) {
```

Comment:
I think in DataSnapshot ctor should be TRI_ASSERT(reader != nullptr);
Otherwise checks like this incorrect

And we should store nullptr to storeSnapshot or use old